requestStorageAccessForOrigin.bs

```diff
@@ -0,0 +1,231 @@
+
+A {{Document}} is in a <dfn>first-party-site context</dfn> if it is the [=active document=] of a [=traversable navigable=]. Otherwise, it is in a [=first-party-site context=] if it is an [=active document=] and the [=environment settings object/origin=] and [=top-level origin=] of its [=relevant settings object=] are [=same site=] with one another.
+
+A {{Document}} is in a <dfn>third party context</dfn> if it is not in a [=first-party-site context=].
+
+<h3 id="the-document-object">Changes to {{Document}}</h3>
+
+<pre class="idl">
+partial interface Document {
+  Promise&lt;undefined> requestStorageAccessForOrigin(USVString origin);
+};
+</pre>
+
+<div algorithm>
+When invoked on {{Document}} |doc| with {{USVString}} |requestedOrigin|, the <dfn export method for=Document><code>requestStorageAccessForOrigin(requestedOrigin)</code></dfn> method must run these steps:
+
+1. Let |p| be [=a new promise=].
+1. If |doc| is not [=Document/fully active=], then [=reject=] |p| with an "{{InvalidStateError}}" {{DOMException}} and return |p|.
+1. If |doc|'s [=node navigable=] is not a [=traversable navigable=], [=reject=] |p| with an "{{InvalidStateError}}" {{DOMException}} and return |p|.
+1. If |doc|'s [=Document/origin=] is an [=opaque origin=], [=reject=] |p| with an "{{InvalidStateError}}" {{DOMException}} and return |p|.
+1. Let |parsedURL| be the the result of running the [=URL parser=] on |requestedOrigin|.
+1. If |parsedURL| is failure, [=reject=] |p| with a "{{TypeError}}" {{DOMException}} and return |p|.
+1. Let |origin| be |parsedURL|'s [=/origin=].
+1. If |origin| is an [=opaque origin=], [=reject=] |p| with an "{{InvalidStateError}}" {{DOMException}} and return |p|.
+1. If |doc|'s [=Document/origin=] is [=same origin=] with |origin|, [=resolve=] and return |p|.
+1. Let |descriptor| be a newly created {{TopLevelStorageAccessPermissionDescriptor}} with {{PermissionDescriptor/name}} set to "<a permission><code>top-level-storage-access</code></a>" and with {{TopLevelStorageAccessPermissionDescriptor/requestedOrigin}} set to |origin|.
+1. Run these steps [=in parallel=]:
+    1. Let |top-level site| be the result of [=obtain a site|obtaining a site=] from |doc|'s [=relevant settings object=]'s [=environment settings object/origin=].
+    1. [=Determine the top-level storage access policy=] with |descriptor|, |doc|, |top-level site|, and |p|.
+1. Return |p|.
+
+</div>
+
+<h3 id="ua-policies">User Agent top-level storage access policies</h3>
+
+<div algorithm>
+To <dfn type="abstract-op">determine if a request has top-level storage access</dfn> with [=request=] |request|, run these steps:
+
+ISSUE: should all cross-site subresource requests behave this way, or only those initiated by the top-level document, or top-level plus those requested from an iframe that is same-site with the top-level document, or only those that have explicitly requested access (in which case a set of requested origins would be maintained)?
```

Comment:
Can we file this issue on the repo and link there?

Reply:
good call, done